apps/dashboard/src/context/environment/environment-provider.tsx

```diff
@@ -35,18 +37,49 @@ function selectEnvironment(environments?: IEnvironment[], selectedEnvironmentId?
 
 export function EnvironmentProvider({ children }: { children: React.ReactNode }) {
   const { currentOrganization } = useAuth();
+  const navigate = useNavigate();
+  const { pathname } = useLocation();
   const [currentEnvironment, setCurrentEnvironment] = useState<IEnvironment | null>(null);
 
+  const switchEnvironmentInternal = useCallback(
+    (allEnvironments?: IEnvironment[], newEnvironmentId?: string) => {
+      const selectedEnvironment = selectEnvironment(allEnvironments, newEnvironmentId);
+      setCurrentEnvironment(selectedEnvironment);
+
+      if (
+        (pathname === ROUTES.ROOT || pathname === ROUTES.ENV || pathname === `${ROUTES.ENV}/`) &&
+        selectedEnvironment
+      ) {
+        navigate(buildRoute(ROUTES.WORKFLOWS, { environmentId: selectedEnvironment._id }));
+        return;
+      } else if (pathname.includes(ROUTES.ENV) && selectedEnvironment) {
+        const newPath = pathname.replace(/\/env\/[^/]+(\/|$)/, `${ROUTES.ENV}/${selectedEnvironment._id}$1`);
+        navigate(newPath);
+        return;
+      }
+
+      navigate(pathname, { replace: true });
+    },
+    [navigate, pathname]
+  );
+
   const { environments, areEnvironmentsInitialLoading } = useFetchEnvironments({
     organizationId: currentOrganization?._id,
-    onSuccess: (environments) => {
-      flushSync(() => setCurrentEnvironment(selectEnvironment(environments, getEnvironmentId())));
-    },
   });
 
-  const switchEnvironment = useCallback(() => {
-    // TODO: in the next PR
-  }, []);
+  useLayoutEffect(() => {
+    if (!environments) {
+      return;
+    }
+
+    const selectedEnvironment = selectEnvironment(environments, getEnvironmentId());
+    switchEnvironmentInternal(environments, selectedEnvironment?._id);
+  }, [environments, switchEnvironmentInternal]);
```

Comment:
when the environment is loaded, or when the component is mounted and we have environments in the query cache --> set the current environment and redirect the user